src/libraries/CalldataDecoder.sol

```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.23;
+
+import {Call} from "./CallLib.sol";
+
+library CalldataDecoder {
+    using CalldataDecoder for bytes;
+    /// @notice equivalent to SliceOutOfBounds.selector, stored in least-significant bits
+
+    error SliceOutOfBounds();
+
+    /// @notice mask used for offsets and lengths to ensure no overflow
+    /// @dev no sane abi encoding will pass in an offset or length greater than type(uint32).max
+    ///      (note that this does deviate from standard solidity behavior and offsets/lengths will
+    ///      be interpreted as mod type(uint32).max which will only impact malicious/buggy callers)
+    uint256 constant OFFSET_OR_LENGTH_MASK = 0xffffffff;
+    uint256 constant OFFSET_OR_LENGTH_MASK_AND_WORD_ALIGN = 0xffffffe0;
+
+    uint256 constant SLICE_ERROR_SELECTOR = 0x3b99b53d;
+
+    /// @notice equivalent to abi.decode(data, (uint256, bytes)) in calldata
+    /// @return nonce The nonce value.
+    /// @return signature The signature to verify.
+    function decodeOpdata(bytes calldata data) internal pure returns (uint256 nonce, bytes calldata signature) {
+        assembly {
+            nonce := calldataload(data.offset)
+        }
+        signature = data.toBytes(1);
+    }
+
+    // TODO length check
+    function decodeCalls(bytes calldata data) internal pure returns (Call[] calldata calls) {
+        assembly {
+            let lengthPtr := add(data.offset, calldataload(data.offset))
+            calls.offset := add(lengthPtr, 0x20)
+            calls.length := calldataload(lengthPtr)
+        }
+    }
+
+    // TODO: Length check
+    function decodeCallsBytes(bytes calldata data)
+        internal
+        pure
+        returns (Call[] calldata calls, bytes calldata opData)
+    {
+        assembly {
+            let lengthPtr := add(data.offset, calldataload(data.offset))
+            calls.offset := add(lengthPtr, 0x20)
+            calls.length := calldataload(lengthPtr)
+
+            let opLengthPtr := add(data.offset, calldataload(add(data.offset, 0x20)))
+            opData.offset := add(opLengthPtr, 0x20)
+            opData.length := calldataload(opLengthPtr)
+        }
+    }
+
+    /// TODO: Import CalldataDecoder.sol, has been previously audited.
```

Comment:
still a TODO or this is the imported code?

Reply:
TODO still, I wanted to see if we could instead import code we've written in CalldataDecorder that we use in v4-periphery but for now I just copied the bottom function in